src/TCommandLine.cpp

```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
 
         }
 
-        switch (ke->key()) {
-        case Qt::Key_Space:
-            if ((ke->modifiers() & (allModifiers & ~(Qt::ShiftModifier))) == Qt::NoModifier) {
-                // Ignore the <SHIFT> modifier only - keeps some users happy!
-                mTabCompletionCount = -1;
-                mAutoCompletionCount = -1;
-                mTabCompletionTyped.clear();
-                if (mpHost->mAutoClearCommandLineAfterSend) {
-                    mHistoryBuffer = -1;
-                } else {
-                    mHistoryBuffer = 0;
-                }
-                mLastCompletion.clear();
-                break;
-
-            } else {
-                // Process as a possible key binding if there are ANY modifiers
-                // other than just a <SHIFT> one; may actaully be configured as
-                // a non-breaking space when used with a modifier!
-                return processPotentialKeyBinding(ke);
+        if (processPotentialKeyBinding(ke)) {
+            return true;
+        }
+        QString keyText;
+        keyText = ke->text(); //default
 
-            }
+        // We want to turn, for example, Control-Up into ESC[1;5A.
+        int modifier = 1 + 1 * (bool(ke->modifiers() & Qt::ShiftModifier)) + 2 * (bool(ke->modifiers() & Qt::AltModifier)) + 4 * (bool(ke->modifiers() & Qt::ControlModifier));
+        QString specialPrefix = QString("\x1b[");
+        if (modifier > 1)
+            specialPrefix = specialPrefix+QString("1;")+('0'+modifier);
 
+        switch (ke->key()) {
+        case Qt::Key_Up:
+            keyText = specialPrefix+"A";
+            break;
+        case Qt::Key_Down:
+            keyText = specialPrefix+"B";
+            break;
+        case Qt::Key_Right:
+            keyText = specialPrefix+"C";
+            break;
+        case Qt::Key_Left:
+            keyText = specialPrefix+"D";
+            break;
+        case Qt::Key_Delete:
+            keyText = specialPrefix+"3~";
+            break;
+        case Qt::Key_Home:
+            keyText = specialPrefix+"H";
+            break;
+        case Qt::Key_End:
+            keyText = specialPrefix+"F";
+            break;
+        case Qt::Key_D:
+            // Readline treats ^D on empty line as EOF and quits.  So don't send it.
+            if ((ke->modifiers() == Qt::ControlModifier) && rl_end == 0)
+                keyText = QString("");
```

Comment:
:book: You do not need to assign an empty string to clear a `QString`:
```
keyText.clear();
```
imparts the same intent and avoids another raw-string to `QString` creation {anyhow the result - an empty `QString` is almost indistinguishable from a null `QString` - only `QString::isNull()` can tell the difference so the recommendation when you *do* want to do an assignment (e.g. in a constructor's initialiser list) is to use `QString()`, though as the default constructor for the QString is the null one anyhow it isn't *needed* in that case}.